### ISSUE TITLE
Support Body Transformations

### ISSUE THREAD

**Issue by arkodg** [envoyproxy/envoy#35783]
*Title*: Support Body Transformations

*Description*:
>Describe the desired behavior, what scenario it enables and how it
would be used.

Envoy supports manipulating/transforming headers, would be great to also support transforming the request and response body to be able to 
* sanitize fields
* support API conversions

[optional *Relevant Links*:]

Adding a list of other proxy implementations that support this
* Amazon API Gateway https://docs.aws.amazon.com/apigateway/latest/developerguide/rest-api-data-transformations.html
* Tyk https://tyk.io/docs/transform-traffic/request-body/
* KrakenD https://www.krakend.io/docs/enterprise/backends/body-generator/
* Gloo Edge https://docs.solo.io/gloo-edge/latest/guides/traffic_management/request_processing/transformations/
* Kong https://docs.konghq.com/hub/kong-inc/request-transformer/
* Apache Apisix https://apisix.apache.org/docs/apisix/plugins/body-transformer/

**Comment by arkodg** [comment]
would be great if this feature/filter can also support copying/setting fields from the body into the header, allowing routing based on request body which is AI LLM Gateway use case, more in [this doc](https://docs.google.com/document/d/1QlpHJwq_GKvLI-WTtuYAzm54B0tfUvzunXYlRcZ-Ap0/edit#heading=h.x4yikowkl073) 
cc @robscott

**Comment by wbpcode** [comment]
Although it is not encouraged that to mutate request body because it will need to buffer whole body and break the streamlined processing. But I also admit that there are lots of related requirements.

So, SGTM. And I can help to review the design and sponsor this new extension if someone wants to take this. (note: a design proposal is necessary first).

**Comment by kyessenov** [comment]
> Although it is not encouraged that to mutate request body because it will need to buffer whole body and break the streamlined processing. But I also admit that there are lots of related requirements.
> 
> So, SGTM. And I can help to review the design and to sponsor this new extension if someone wants to take this. (note: a design proposal is necessary first).

I think the technical challenge is to make mutations streaming. It's feasible as long as the body is structured. The buffering approach forces large connection buffers in multiplexed protocols, which is not scaleable for multi-tenant gateways.

**Comment by wbpcode** [comment]
> > Although it is not encouraged that to mutate request body because it will need to buffer whole body and break the streamlined processing. But I also admit that there are lots of related requirements.
> > 
> > So, SGTM. And I can help to review the design and to sponsor this new extension if someone wants to take this. (note: a design proposal is necessary first).
> 
> I think the technical challenge is to make mutations streaming. It's feasible as long as the body is structured. The buffering approach forces large connection buffers in multiplexed protocols, which is not scaleable for multi-tenant gateways.

According to my exp, in the scenarios where this feature is required, the body basically is a JSON.
It's almost impossible to make mutations streaming for that.

At least for now, I don't know how to make a **general solution** for long live stream which has unlimited body length. So, I am inclined to ignore them first.

**Comment by arkodg** [comment]
+1 to @wbpcode's suggestion of keep streaming out of scope

**Comment by kyessenov** [comment]
> +1 to @wbpcode's suggestion of keeping streaming, out of scope

Yes, I didn't mean to require it. But we should be explicit about the limitations of a buffering approach (e.g. connection buffers must be at least the number of streams per connections \times max buffered bytes). E.g. 1MB JSONs with 100 H2 streams require up to 100MB connection buffers.

**Comment by mathetake** [comment]
https://docs.google.com/document/d/1odMAistdE8OrJHqKvV4ou9vsZxB1Wepy0cxAyBJ-Bes/edit so @arkodg and I quickly wrote a simple proposal. @wbpcode could you take a look when you get a chance? Thanks in advance!

**Comment by kyessenov** [comment]
CC @TAOXUY - I think you also need some body transformations and extractions.

Back to my early buffering point - I think it would be nice to have separate buffering and streaming modes. You don't need to buffer to sanitize JSON fields, and the only meaningful benefit of this over ext_proc is that it must perform significantly better (e.g. hundred of micros P50). overhead, which is hard to do when you fully buffer.

**Comment by TAOXUY** [comment]
Yeah, our case is extracting protobuf and our needs are satisfied by these 2 filters https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/grpc_field_extraction_filter
and https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/proto_message_extraction_filter

**Comment by mathetake** [comment]
@wbpcode could you take a look at the doc we shared above when you get a change? tia!

**Comment by wbpcode** [comment]
I get some free bandwidth and will give this a try to support the substitution formatter based body transformation. At the initial version, it will support following feature:

1. transform request and response body based on the https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/substitution_format_string.proto#envoy-v3-api-msg-config-core-v3-jsonformatoptions.
2. request/response headers mutations based on the body content.
3. filter state mutations based on the body content. (to extract data from body to filter state for logging/stats/other filters but not expose to clients)
4. Only support JSON request and JSON response. (will support stream/event in the future.)
